### ISSUE TITLE
what does test_minibatch exactly return

### ISSUE THREAD

**Issue by JasonRuan5** [Microsoft/CNTK#2461]
What does exactly trainer.test_minibatch(data) returns?  The document says the average evaluation criterion value per sample for the tested minibatch, but what is the evaluation criterion? I was told it returns the average value of the loss and also told it returns the ratio of miss-classified count and test sample count. I have created a python program based on TrainResNet_CIFAR10.py to train and test about 15000 and 3500 of (32x32x3) images with 4 classes. As test, we have extracted the first 4 out of 10 classes from the existing train_map and test_map so they contains 20000 train and 4000 test images. I've created two methods below in hope to clarify this.

errTest = calc_metric(trainer, reader_test, test_epoch_size, label_var, input_map)
lossEval, errEval  = eval_metric(z, testList, mean_bgr)

    def calc_metric(trainer, reader_test, test_epoch_size, label_var, input_map) :
        # Evaluation parameters
        minibatch_size = 16

        # process minibatches and evaluate the model
        metric_numer    = 0
        metric_denom    = 0
        sample_count    = 0

        while sample_count < test_epoch_size:
            current_minibatch = min(minibatch_size, test_epoch_size - sample_count)
            # Fetch next test min batch.
            data = reader_test.next_minibatch(current_minibatch, input_map=input_map)
            # minibatch data to be trained with
            metric_numer += trainer.test_minibatch(data) * current_minibatch
            metric_denom += current_minibatch
            # Keep track of the number of samples processed so far.
            sample_count += data[label_var].num_samples

        return (metric_numer / metric_denom)

    def eval_metric(z, fileList, mean_value) :
        test_size = len(fileList)

        ce = 0.0  #average loss
        cm = 0.0  #ratio of miss classified and total test sample
        n = 0
        while n < test_size:
            pred = calc_pred(z, fileList[n].filename, mean_value)
            pred = softmax(pred)
            ce += - np.log(pred[fileList[n].classID]) 
            top_class = np.argmax(pred)
            if (top_class != fileList[n].classID) :
                cm += 1
            n += 1

        return (ce / test_size, cm / test_size)

    def calc_pred(z, img_name, mean_value) :
        rgb_image = np.asarray(Image.open(img_name), dtype=np.float32)
        bgr_image = rgb_image[..., [2, 1, 0]] - mean_value
        pic = np.ascontiguousarray(np.rollaxis(bgr_image, 2))
        probs = z.eval({z.arguments[0]:[pic]})
        predictions = np.squeeze(probs)
        return predictions

    def softmax(x):
        return np.exp(x) / np.sum(np.exp(x), axis=0)

I found errEval is closer to errTest than lossEval. With the sample images, errTest is about 5% less than errEval consistently and all three decrease with the train loss up to 100 epochs. With our own images, errEval starts with over 50% higher than errTest, while errTest decreases with train loss as sample images, but errEval starts to increases after about a dozen epochs and at end of 160 epoch, it becomes 3 time larger than errTest.

Our images do not have equal number of images for each class -- the first class may have twice as second, 4 time as third, 8 time as the last class. Our images are individual nuclei cells, which may not be as easy to be trained as sample images. My questions are: 
1. What does test_minibatch exactly returns?    
2. Why the two functions return very much the same (for errTest and errEval) with the sample images, but very different behavior for our own images? 
3. What could be the root cause of the problem and any possible solutions?

**Comment by eldakms** [comment]
1. test_minibatch simply runs Forward path on the current network and calculates the evaluation function (not the loss one!) that you provided to the trainer (second in the tuple, not the first one, see here https://www.cntk.ai/pythondocs/cntk.train.trainer.html):
_criterion (tuple of Function or Variable) - Function with one or two outputs, representing loss and, if given, evaluation metric (in this order). Alternatively, a tuple(loss Function, evaluation Function) is also accepted._
It averages the result of the evaluation function over the number of samples in the minibatch and returns this averaged value, in pseudo code:
test_minibatch([x1, ... xN]) = sum(eval(x1) + ... eval(xN))/N

2. Are you sure you input is the same? How have you instantiated the test reader? Does it do any transformations? How did you instantiated the trainer, did you give the equivalent evaluation function that you are calculating in eval_metric?

3. function::eval and trainer::test_minibatch both internally use forward, there should be no different between them except for averaging

**Comment by JasonRuan5** [comment]
1. Great and it makes sense! 
    pe = classification_error(z, label_var) is used as second argument in 
    trainer = Trainer(z, (ce, pe), learner, progress_writers).  
The test_minibatch should return the ratio of miss-classified count and test sample count, right?
2. How do we verify input is the same?  Can you tell from def calc_pred(z, img_name, mean_value) above if input is the same? I do worry if it is correct or not, for instance, should mean_value be RGB or BGR order and should mean_value subtract from rgb_image or bgr_image as below?
  rgb_image = np.asarray(Image.open(img_name), dtype=np.float32)
  bgr_image = rgb_image[..., [2, 1, 0]] - mean_value

    We used the standard transformation when create_reader as below:
        transforms += [ xforms.scale(width=image_width, height=image_height, 
        channels=num_channels, interpolations='linear'), xforms.mean(mean_file) ]
    What will xforms.scale and particularly xforms.mean do?
3. Great to know! So far, errTest gives us very encouraging ratio. If we could make errEval gives the similar value, which is what we got when used in C#, it would be super.   
The python code is based on TrainResNet_CIFAR10.py, Is it possible to post here or email?  Thanks so much!

**Comment by eldakms** [comment]
1. Right
2. We use OpenCV in the readers, effectively it reads in BGR (using opencv), performs scale with interpolation, subtracts the mean values, does HWC->CHW transformation. The last one is important, because CuDNN/CNTK only works with CHW. It is done transparently by the MinibatchSource. It seems some of the steps are missing in your manually written function.

The detailed code of transformations if you are interested is here (Scale + Mean, https://github.com/Microsoft/CNTK/blob/master/Source/Readers/ImageReader/ImageTransformers.cpp)

You can have a look _test_base64_image_deserializer_ test in the unit tests (https://github.com/Microsoft/CNTK/blob/master/bindings/python/cntk/io/tests/io_tests.py), there we read from the deserializer and comparing it to the manually read file.

I would say if you want to make sure the input is the same - just dump the output of the minibatch with one image and then make sure that on the same image the output of your function matches this binary representation.

3. I am not on CNTK team, but they should help you if you send an email to Cha (or post your stuff here)

**Comment by JasonRuan5** [comment]
Thanks so much.  I do suspect there might be problem with def calc_pred(z, img_name, mean_value).  As you suggested, I’ve tried to compare 
    pic = np.ascontiguousarray(np.rollaxis(bgr_image, 2)) 
   data = reader_test.next_minibatch(current_minibatch, input_map=input_map) 

and make sure the input is the same.  We can see what values that pic contains, but cannot see data values.  Is there any way to view what data contains in quick view? We have changed mean_value from HWC order to CHW order and revised calc_pred as below

    def calc_pred(z, img_name, mean_chw) :
        rgb_image = np.asarray(Image.open(img_name), dtype=np.float32)
        bgr_image = rgb_image[..., [2, 1, 0]]
        bgr_chw = np.rollaxis(bgr_image, 2)
        pic = np.ascontiguousarray(bgr_chw - mean_chw)
        probs = z.eval({z.arguments[0]:[pic]})
        predictions = np.squeeze(probs)
        return predictions

but still get the similar results as before. Is there any problem with the above newly revised calc_pred method?

**Comment by JasonRuan5** [comment]
Just found out if we comment out xforms.mean(mean_file) in the test reader and without subtracting mean_chw from brg_chw in calc_pred

    transforms += [
        xforms.scale(width=image_width, height=image_height, channels=num_channels,   
    interpolations='linear'),
        xforms.mean(mean_file)
    ]
    pic = np.ascontiguousarray(bgr_chw - mean_chw)
errTest and errEval are identical.  Now, the question is what xforms.mean(mean_file) does and how we should subtract mean_chw?

**Comment by eldakms** [comment]
it just subtracts the mean:
```
if (m_meanImg.size() == mat.size())
    {
        // If matrix has not been converted to the right type, do it now as mean requires floating point type.
        ConvertToFloatingPointIfRequired(mat);
        mat = mat - m_meanImg;
    }
    else
    {
        fprintf(stderr, "WARNING: Mean file does not match the size of the input image, will be ignored.\n"
            "Please remove mean transformation from the config.\n");
    }
```

where m_meanImg is the mean file:
```
       cv::FileStorage fs;
       fs.open(msra::strfun::utf8(meanFile).c_str(), cv::FileStorage::READ);
        if (!fs.isOpened())
            RuntimeError("Could not open file: %ls", meanFile.c_str());
        fs["MeanImg"] >> m_meanImg;
        int cchan;
        fs["Channel"] >> cchan;
        int crow;
        fs["Row"] >> crow;
        int ccol;
        fs["Col"] >> ccol;
        if (cchan * crow * ccol !=
            m_meanImg.channels() * m_meanImg.rows * m_meanImg.cols)
            RuntimeError("Invalid data in file: %ls", meanFile.c_str());
        fs.release();
        m_meanImg = m_meanImg.reshape(cchan, crow);
```

**Comment by JasonRuan5** [comment]
We have traced into MeanTransformer::MeanTransformer and Apply methods and found m_meanImg.data contains exactly the same data as mean_chw when mat = mat - m_meanImg is executed.  However, pic and data below

    pic = np.ascontiguousarray(bgr_chw - mean_chw)
    data = reader_test.next_minibatch(current_minibatch, input_map=input_map)

are different.  Only the first float point is the same.  I thought every bgr value in chw order is subtracted by their corresponding mean value, then why pic and data is different?  I did checked, if no mean is applied, pic and data are exactly the same, which means it much be caused by how mean is applied.  I am really puzzled.  Any advices or clues would be greatly appreciated.

**Comment by eldakms** [comment]
are you sure your mean_chm and meanfile is the same? I think mean file is in HWC because it does transformation using OpenCV (which is HWC).  can you dump m_meanImg and compare binary wise with mean_chw?

Thanks!

**Comment by vijtad** [comment]
Sent from my T-Mobile 4G LTE Device
-------- Original message --------From: JasonRuan5 <notifications@github.com> Date: 10/17/17  11:13 PM  (GMT-05:00) To: Microsoft/CNTK <CNTK@noreply.github.com> Cc: Subscribed <subscribed@noreply.github.com> Subject: Re: [Microsoft/CNTK] what does test_minibatch exactly return (#2461) 
Just found out if we comment out xforms.mean(mean_file) in the test reader and without subtracting mean_chw from brg_chw in calc_pred
transforms += [
    xforms.scale(width=image_width, height=image_height, channels=num_channels,   
interpolations='linear'),
    xforms.mean(mean_file)
]
pic = np.ascontiguousarray(bgr_chw - mean_chw)

errTest and errEval are identical.  Now, the question is what xforms.mean(mean_file) does and how we should subtract mean_chw?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/Microsoft/CNTK","title":"Microsoft/CNTK","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/Microsoft/CNTK"}},"updates":{"snippets":[{"icon":"PERSON","message":"@JasonRuan5 in #2461: Just found out if we comment out xforms.mean(mean_file) in the test reader and without subtracting mean_chw from brg_chw in calc_pred\r\n\r\n    transforms += [\r\n        xforms.scale(width=image_width, height=image_height, channels=num_channels,   \r\n    interpolations='linear'),\r\n        xforms.mean(mean_file)\r\n    ]\r\n    pic = np.ascontiguousarray(bgr_chw - mean_chw)\r\nerrTest and errEval are identical.  Now, the question is what xforms.mean(mean_file) does and how we should subtract mean_chw?\r\n"}],"action":{"name":"View Issue","url":"https://github.com/Microsoft/CNTK/issues/2461#issuecomment-337448211"}}}

**Comment by JasonRuan5** [comment]
@eldakms Thanks so much!  That could explain everything.  We thought the mean_file is in chw order.  If in hwc order, is it in bgr or rgb order?  We will revise accordingly and hope that will fix the problem.

**Comment by eldakms** [comment]
OpenCV is bgr if i'm not mistaken, so it should be bgr.

**Comment by JasonRuan5** [comment]
I’ve just changed mean_chw to mean_bgr and revised calc_pred as below:

    def calc_pred(z, img_name, mean_bgr) :
        rgb_image = np.asarray(Image.open(img_name), dtype=np.float32)
        bgr_image = rgb_image[..., [2, 1, 0]]  - mean_bgr
        bgr_chw = np.rollaxis(bgr_image, 2)
        pic = np.ascontiguousarray(bgr_chw)
        probs = z.eval({z.arguments[0]:[pic]})
        predictions = np.squeeze(probs)
        return predictions

That indeed fixed the problem.  Thanks so much for your help!
